File: 源码/bulubulu/module_c_func/func_base/src/main/java/cn/kuwo/pp/http/RetrofitClient.java
package cn.kuwo.pp.http;

import android.util.Log;


import java.io.File;
import java.net.ConnectException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import cn.kuwo.common.app.App;
import cn.kuwo.common.utilscode.UtilsCode;
import cn.kuwo.networker.BaseInterceptor;
import cn.kuwo.networker.CacheInterceptor;
import cn.kuwo.networker.EncryptInterceptor;
import cn.kuwo.networker.SchedulerTransformer;
import cn.kuwo.networker.Urls;
import cn.kuwo.pp.http.bean.BaseHttpResult;
import cn.kuwo.networker.exception.ApiException;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.functions.Function;
import me.jessyan.progressmanager.ProgressManager;
import okhttp3.Cache;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;


/**
 * Created by shihc on 2016/10/31.
 */

public class RetrofitClient {
    private Retrofit mRetrofit;
    private ApiService mApiService;

    //在访问HttpMethods时创建单例
    private static class SingletonHolder {
        private static final RetrofitClient INSTANCE = new RetrofitClient();
    }

    //获取单例
    public static RetrofitClient getInstance() {
        return SingletonHolder.INSTANCE;
    }

    public RetrofitClient() {
        mRetrofit = new Retrofit.Builder()
                .baseUrl(Urls.URL_RELEASE)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .client(getHttpClient())
                .build();
        mApiService = mRetrofit.create(ApiService.class);
    }

    public static ApiService getApiService() {
        return getInstance().mApiService;
    }

    //设置OkHttpClient参数，拦截器、超时等
    private OkHttpClient getHttpClient() {
        //缓存容量
        long SIZE_OF_CACHE = 100 * 1024 * 1024; // 100 MiB
        File cacheDir = App.getInstance().getCacheDir();
        if (App.DEBUG) Log.d("cacheDir", cacheDir.getAbsolutePath());
        Cache cache = new Cache(new File(cacheDir, "responses"), SIZE_OF_CACHE);
        //设置日志拦截器
        OkHttpClient.Builder builder = ProgressManager.getInstance().with(new OkHttpClient.Builder());// 构建 OkHttpClient 时,将 OkHttpClient.Builder() 传入 with() 方法,进行初始化配置
        builder.addInterceptor(getHttpLoggingInterceptor());
        builder.addNetworkInterceptor(new CacheInterceptor());
        builder.addInterceptor(new BaseInterceptor());
        builder.addInterceptor(new EncryptInterceptor());
        builder.connectTimeout(20, TimeUnit.SECONDS);
        builder.readTimeout(40, TimeUnit.SECONDS);
        builder.writeTimeout(40, TimeUnit.SECONDS);
        builder.cache(cache);
        builder.retryOnConnectionFailure(true);

        return builder.build();
    }

    //请求头中的参数，公共也放到这里
    private Map<String, String> getHeaderParams() {
        Map<String, String> map = new HashMap<>();
        return map;
    }

    //日志拦截器，打印请求日志
    private HttpLoggingInterceptor getHttpLoggingInterceptor() {
        HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
        loggingInterceptor.setLevel(App.DEBUG ? HttpLoggingInterceptor.Level.BODY : HttpLoggingInterceptor.Level.NONE);
        return loggingInterceptor;
    }

    public void execute(Observable observable, Observer observer) {
        toSubscribe(observable, observer, true);
    }

    public void executeWithoutBaseBean(Observable observable, Observer observer) {
        toSubscribe(observable, observer, false);
    }

    /**
     * 发起请求并处理返回数据
     *
     * @param observable    ApiService 定义
     * @param observer      请求回调
     * @param parseBaseBean 是否解析 BaseHttpResult
     * @param <T>
     */
    private <T> void toSubscribe(Observable<T> observable, final Observer<T> observer, boolean parseBaseBean) {
        if (observer != null && !(observer instanceof CustomObserver)) {
            throw new RuntimeException("observer must use CustomObserver");
        }
        if (parseBaseBean) {
            observable = observable.map(new ServerResponseFunc());
        }
        observable.onTerminateDetach().compose(new SchedulerTransformer<>()).subscribe(observer);
    }

    /**
     * 用来统一处理Http的resultCode,并将HttpResult的Data部分剥离出来返回给subscriber
     *
     * @param <T> Subscriber真正需要的数据类型，也就是Data部分的数据类型
     */
    public static class ServerResponseFunc<T> implements Function<BaseHttpResult<T>, T> {

        @Override
        public T apply(BaseHttpResult<T> httpResult) throws Exception {
            if (httpResult.isSuccess()) {
                T data = httpResult.getData();
                if (data == null) {
                    throw new ApiException(null, httpResult.getStatus(), "返回数据为空");
                }
                return data;
            }
            if (!UtilsCode.INSTANCE.isConnected()) {
                throw new ConnectException("网络不可用");
            } else {
                throw new ApiException(null, httpResult.getStatus(), httpResult.getMsg());
            }
        }
    }

}

File: 源码/bulubulu/module_r_library/networker/src/main/java/cn/kuwo/networker/RetrofitClient.kt
package cn.kuwo.networker

import android.text.TextUtils
import cn.kuwo.common.app.App
import me.jessyan.progressmanager.ProgressManager
import okhttp3.Cache
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Response
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.io.File
import java.util.concurrent.TimeUnit

class RetrofitClient private constructor() {


    companion object {
        // 双重校验锁
        private val instance: RetrofitClient by lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED) { RetrofitClient() }

        fun <T> api(service: Class<T>): T {
            return instance.api(service)
        }
    }


    private val mOkHttpClient: OkHttpClient =
        ProgressManager.getInstance().with(OkHttpClient.Builder())
            .addInterceptor(HttpLoggingInterceptor().setLevel(if (BuildConfig.DEBUG) HttpLoggingInterceptor.Level.BODY else HttpLoggingInterceptor.Level.NONE))
            .addNetworkInterceptor(CacheInterceptor())
            .addInterceptor(BaseInterceptor())
            .addInterceptor(EncryptInterceptor())
            .connectTimeout(20, TimeUnit.SECONDS)
            .readTimeout(40, TimeUnit.SECONDS)
            .writeTimeout(40, TimeUnit.SECONDS)
            .cache(Cache(File(App.getInstance().cacheDir, ""), 104_857_600)) //104_857_600 100M
            .build()


    private val mRetrofit: Retrofit = Retrofit.Builder()
        .baseUrl("")
        .addConverterFactory(GsonConverterFactory.create())
        .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
        .client(mOkHttpClient)
        .build()


    private val mapApi: HashMap<String, Any?> = hashMapOf()

    // 确保每个返回的都是正确的
    fun <T> api(service: Class<T>): T {
        // 先看看map中是否有缓存
        var get = mapApi[service.simpleName] as? T
        if (get != null) {
            get = mRetrofit.create(service)//instance.create(service)
            // 不允许为空
            mapApi[service.simpleName] = get
        }
        return get ?: throw NullPointerException("创造的接口不允许为空！")
    }


    interface ApiTest {
        fun test()
    }

    private fun test() {
        val api = RetrofitClient.instance.api(ApiTest::class.java) // 可
        api.test()
    }


    class CacheInterceptor : Interceptor {
        override fun intercept(chain: Interceptor.Chain): Response {
            val request = chain.request()
            val response = chain.proceed(request)
            val serverCache = response.header("Cache-Control")
            // 如果服务端设置相应的缓存策略那么遵从服务端的不做修改
            // 如果服务端设置相应的缓存策略那么遵从服务端的不做修改
            if (TextUtils.isEmpty(serverCache)) {
                val cacheControl = request.cacheControl().toString()
                if (!TextUtils.isEmpty(cacheControl)) {
                    return response.newBuilder()
                        .addHeader("Cache-Control", cacheControl)
                        .removeHeader("Pragma")
                        .build()
                }
            }
            return response
        }
    }

    class BaseInterceptor : Interceptor {
        override fun intercept(chain: Interceptor.Chain): Response {
            val request = chain.request()
//        if (TextUtils.equals("false", request.header("addBaseParameter"))) {
//            return chain.proceed(request);
//        }
//        //添加公共参数
//        HttpUrl httpUrl = request.url();
//        HttpUrl.Builder builder = httpUrl.newBuilder();
//        if (!TextUtils.isEmpty(UserInfoManager.INSTANCE.getUid())){
//            builder.addQueryParameter("uid", UserInfoManager.INSTANCE.getUid());
//            builder.addQueryParameter("sid", UserInfoManager.INSTANCE.getToken());
//        }else{
//            builder.addQueryParameter("uid", "0");
//            builder.addQueryParameter("sid", "0");
//        }
//        if (UserInfoManager.INSTANCE.isLogin()){
//            builder.addQueryParameter("sex", UserInfoManager.INSTANCE.getUserGender());
//        } else {
//            builder.addQueryParameter("sex", String.valueOf(MMKV.defaultMMKV().getInt("gender", 1)));
//        }
//        Request.Builder requestBuilder = request.newBuilder().url(builder.build());
//        requestBuilder.addHeader("devId",DeviceUtils.getAndroidID());
//        requestBuilder.addHeader("appUid", PPApp.getAppUid());
//        requestBuilder.addHeader("token", UserInfoManager.INSTANCE.getToken());
//        requestBuilder.addHeader("plat","android");
//        requestBuilder.addHeader("from","ar");
//        requestBuilder.addHeader("src","kuwo");
//        requestBuilder.addHeader("version",String.valueOf(AppUtils.getAppVersionCode()));
//        requestBuilder.addHeader("api-ver","1");
//        Request request1 = requestBuilder.build();
//        LogUtils.d(request1.headers().toString());
            return chain.proceed(request)
        }

    }

    class EncryptInterceptor : Interceptor {
        override fun intercept(chain: Interceptor.Chain): Response {
            val request = chain.request()
            /*
                    Request request = chain.request();
        if (TextUtils.equals("true", request.header("encrypt"))) {
            HttpUrl oldHttpUrl = request.url();
            HttpUrl newUrl = oldHttpUrl.newBuilder().build();
            String url = newUrl.url().toString();
            LogUtils.d("--> " + url);
            String timeValue = String.valueOf(System.currentTimeMillis());
            String encodeParams = AccountUrlUtil.encryptJsonParams(url.substring(url.indexOf("?") + 1));
            LogUtils.d(encodeParams);
            String sign = MD5.getMD5Str("bulubulu"+encodeParams+timeValue);
            newUrl = newUrl.newBuilder().encodedQuery(null)
                    .addQueryParameter("sign", sign.toUpperCase())
                    .addQueryParameter("appId", "bulubulu")
                    .addQueryParameter("time", timeValue)
                    .addQueryParameter("data", encodeParams).build();
            Request.Builder requestBuilder = request.newBuilder().url(newUrl);
            return chain.proceed(requestBuilder.build());
        }else{
            return chain.proceed(request);
        }
             */
            return chain.proceed(request)
        }

    }


}